### ISSUE TITLE
Generate 'vwmacc' instructions for quantized GEMM

### ISSUE THREAD

**Issue by dcaballe** [iree-org/iree#10276]
We currently generate the following instruction sequence for a i8/i32 GEMMs:

```
     vsetvli   zero,a5,e32,m2,ta,mu
     vle8.v    v26,(s5)
     vsext.vf4 v28,v26
     vmacc.vx  v24,s8,v28
     vmacc.vx  v22,s9,v28
     vmacc.vx  v20,s10,v28
     vmacc.vx  v18,s11,v28
     vmacc.vx  v16,ra,v28
     vmacc.vx  v14,a6,v28
     vmacc.vx  v12,a7,v28
     vmacc.vx  v10,t0,v28
```

These sequence can be improved by using `vwmacc` instead of `vmacc` (and consequently replacing `vsext.vf4` with `vsext.vf2`). The RISC-V community pointed out that the best solution is to modify one of the ISel combiners that already takes care of this replacement. In particular, the combiner is [here](https://github.com/llvm/llvm-project/blob/main/llvm/lib/Target/RISCV/RISCVISelLowering.cpp#L8764) and it converts `vmul` into a `vwmul` (building block to later generate the intended `vwmacc`) only when their operand is a `vsext` *with a single use*. The idea would be to extend the combiner so that `vsext` can have multiple uses that we know are going to be turned into a `vwmul` as well.

Perhaps Quentin could provide some feedback!

**Comment by dcaballe** [comment]
Repro: `llc -mtriple=riscv64 -mattr=+v -verify-machineinstrs < repro.ll`

```
; Supported:
;       vsetvli a0, zero, e8, mf4, ta, mu
;       vwmul.vv  v10, v8, v9
define <vscale x 2 x i16> @single_vwmul_i8_i16(<vscale x 2 x i8> %va, <vscale x 2 x i8> %vb) {
  %vc = sext <vscale x 2 x i8> %va to <vscale x 2 x i16>
  %vd = sext <vscale x 2 x i8> %vb to <vscale x 2 x i16>
  %ve = mul <vscale x 2 x i16> %vc, %vd
  ret <vscale x 2 x i16> %ve
}
                                                                                                                                                                                      
; Not supported:
;       vsext.vf2
;       vsext.vf2
;       vwmul.vv
define <vscale x 2 x i32> @single_vwmul_i8_i32(<vscale x 2 x i8> %va, <vscale x 2 x i8> %vb) {
  %vc = sext <vscale x 2 x i8> %va to <vscale x 2 x i32>
  %vd = sext <vscale x 2 x i8> %vb to <vscale x 2 x i32>
  %ve = mul <vscale x 2 x i32> %vc, %vd
  ret <vscale x 2 x i32> %ve
}

; Not supported:
;       vsext.vf2
;       vsext.vf2
;       vsext.vf2
;       vwmul.vv
;       vwmul.vv
define <vscale x 2 x i32> @multi_vwmul(<vscale x 2 x i8> %va, <vscale x 2 x i8> %vb, <vscale x 2 x i8> %vc) {
  %vd = sext <vscale x 2 x i8> %va to <vscale x 2 x i32>
  %ve = sext <vscale x 2 x i8> %vb to <vscale x 2 x i32>
  %vf = sext <vscale x 2 x i8> %vc to <vscale x 2 x i32>
  %vg = mul <vscale x 2 x i32> %vd, %ve
  %vh = mul <vscale x 2 x i32> %vd, %vf
  %vi = or <vscale x 2 x i32> %vg, %vh
  ret <vscale x 2 x i32> %vi
} 
```

The unsupported cases still need `vsext` instructions but it reduces the number of registers needed so that should give us more budget for unrolling and more aggressive vectorization. I think we should hack something quickly to measure the performance impact, to make sure the transformation is actually worth it.

**Comment by dcaballe** [comment]
Assigning this to @qcolombet for now so that he can provide some feedback when he has the time.

**Comment by qcolombet** [comment]
Thanks @dcaballe for the heads-up.
Taking a look.

**Comment by qcolombet** [comment]
> The idea would be to extend the combiner so that vsext can have multiple uses that we know are going to be turned into a vwmul as well.

I'm attaching two patches in that direction (github doesn't like the `.patch` extension, so I appended a `.txt`):
1. `fold_aggressively` skips the "one use" check all together and applies the optimization (vmul(ext) -> vwmul) aggressively
2. `fold_when_other_ops_may_fold` checks that the other users of the extension are VW candidates and if that's the case, allow the optimization to proceed.

#1 is not as crazy as it may sound, depending on the micro architecture, keeping the extension while allowing the folding may be fine if the folding in the target instruction is free. There may be a register pressure cost though (we would keep both the source of the extension and its destination alive at the same time.)

#2 suffers exactly the same problems: if the folding doesn't happen for all the instructions, we'll have the same situation with both the source and destination of the extension alive at the same time.

Finally, because of the SDISel framework, none of the proposed patches (as well as the existing code to be fair) would fold if the use is in a different basic block than the extension. To fix that, RISC-V would need to switch to GlobalISel ;).

@dcaballe give a try to the patches and let me know how it goes.
Then we can work with the RISC-V people to see how they want to properly do this.

[fold_aggressively.patch.txt](https://github.com/iree-org/iree/files/9522041/fold_aggressively.patch.txt)
[fold_when_other_ops_may_fold.patch.txt](https://github.com/iree-org/iree/files/9522042/fold_when_other_ops_may_fold.patch.txt)

**Comment by dcaballe** [comment]
Awesome! Both implementations lead to 18% and 12% improvement on MobileBERT-quant and PersonDetect, respectively.

I think #2 is a good starting point! A few comments:
  - There is another `hasOneUse` in the combiner that we may want to address. 
  - I'm also wondering if there would be a way to pre-compute if an op can fold before actually folding it. In that way, we would be certain that everything will fold or bail out, otherwise. 

> keeping the extension while allowing the folding may be fine if the folding in the target instruction is free. There may be a register pressure cost though (we would keep both the source of the extension and its destination alive at the same time.)

Note that it's not about removing or keeping the extension. For the i8->i32 case, the original `vsext.vf4` turns into a `vsext.vf2` when folding so if folding doesn't happen for any of the users we would end up with two extensions: `vsext.vf2` and `vsext.vf4`. I guess that would be slower or faster than not folding depending on how many ops are folded and register pressure. Taking into account that folding should reduce the register pressure, it might always be worth folding.

Feel free to proceed with this as you want! Happy if you want to post this on phabricator. Just make sure you include Craig. He is the one that gave me all the initial feedback.

Thanks a lot for looking into this!

**Comment by qcolombet** [comment]
> I'm also wondering if there would be a way to pre-compute if an op can fold before actually folding it. In that way, we would be certain that everything will fold or bail out, otherwise.

TL;DR, I think that's doable, but it requires some work.

I have to look closer, IIRC it's possible to do that, but that kind of transformations doesn't fit nicely in SDISel framework. SDISel usually expects that you replace the root of the pattern, whereas here we would need to go backward w.r.t. that scheme (i.e., SDISel likes combining an op with its args, whereas here we would combine an op with its uses.)

>  [...]if folding doesn't happen for any of the users we would end up with two extensions

Good point!
@dcaballe Actually now that I look again at the assembly, why is the code faster?

We replace one extension with another one (`vsext.vf4` vs. `vsext.vf2`) and one `mul` with another one (`vmul` vs. `vwmul`).

**Comment by dcaballe** [comment]
> I have to look closer, IIRC it's possible to do that, but that kind of transformations doesn't fit nicely in SDISel framework.

I think it's worth having this discussion with the RISC-V community. We can use your patch and see how it goes.

> @dcaballe Actually now that I look again at the assembly, why is the code faster?
>
> We replace one extension with another one (vsext.vf4 vs. vsext.vf2) and one mul with another one (vmul vs. vwmul).

`vsext.vf4` extends the `macc` operands to `i32` whereas `vsext.vf2` extends them to `i16`. That means that we use 1/2 less registers (lower IMUL) when `vsext.vf2` is used and that should impact the latency of the `macc` ops.

**Comment by qcolombet** [comment]
> vsext.vf4 extends the macc operands to i32 whereas vsext.vf2 extends them to i16. That means that we use 1/2 less registers (lower IMUL) 

Ah makes sense.
I missed that because the registers have the same name irrespective to the width of the data.

**Comment by qcolombet** [comment]
Posted a slightly improved version of the #2 patch at https://reviews.llvm.org/D133739 to get the conversation started with Craig.

**Comment by qcolombet** [comment]
Forgot to post the update here.
I've refactored the code SDISel to make the folding decisions centralized https://reviews.llvm.org/D134703.
Then I've updated https://reviews.llvm.org/D133739 for a slightly better term solution.

Long term, I believe the best way to fix that would be in a separate machine pass, ideally as part of instruction selection since it would allow to compose/reuse the other combines (like `vwmacc`), but that would require GISel and RISC-V doesn't support it (yet).

**Comment by dcaballe** [comment]
Thanks @qcolombet for pushing this forward! I think we should find a path to the longer term approach. I was looking into the RVV spec and, in addition to adds, subs, muls and their fused counterparts, we also have widening reductions so a similar folding might be needed in the future for them. Let's see if we hear back from the community.

**Comment by dcaballe** [comment]
I can reproduce performance after integration. Great!
